asf_welsh_energy_consultation/pipeline/process_data.py

```diff
@@ -0,0 +1,256 @@
+        .size()
+        .reset_index()
+        .pivot(index="year", columns="HP_INSTALLED")
+    )
+
+    new_hp_counts.columns = new_hp_counts.columns.droplevel(0)
+    new_hp_counts.columns.name = None
+    new_hp_counts = new_hp_counts.rename(columns={False: "Other", True: "Heat pump"})
+
+    new_hp_counts["n"] = new_hp_counts.sum(axis=1)
+    new_hp_counts["prop_hp"] = new_hp_counts["Heat pump"] / new_hp_counts["n"]
+
+    new_builds = new_hp_counts.reset_index()
+
+    new_builds = pd.melt(new_builds, id_vars="year", value_vars=["Other", "Heat pump"])
+    new_builds["year"] = pd.to_datetime(new_builds["year"], format="%Y")
+
+    new_builds = new_builds.rename(columns={"variable": "Heating system"})
+
+    return new_builds
+
+
+def get_new_hp_cumsums():
+    """Get cumulative total of new build HPs.
+
+    Returns:
+        pd.DataFrame: New build HPs cumulative totals.
+    """
+    wales_epc_new = get_wales_epc_new()
+
+    new_hps = wales_epc_new.loc[wales_epc_new["HP_INSTALLED"]].reset_index(drop=True)
+    new_hps_sums = (
+        new_hps.sort_values("INSPECTION_DATE")
+        .groupby("INSPECTION_DATE")["HP_INSTALLED"]
+        .sum()
+    )
+
+    idx = pd.date_range(new_hps_sums.index.min(), new_hps_sums.index.max())
+    new_hps_sums.index = pd.DatetimeIndex(new_hps_sums.index)
+    new_hps_sums = new_hps_sums.reindex(idx, fill_value=0)
+    new_hps_cumsums = (
+        new_hps_sums.cumsum()
+        .reset_index()
+        .rename(columns={"index": "Date", "HP_INSTALLED": "Number of EPCs"})
+    )
+
+    return new_hps_cumsums
+
+
+def mcs_epc_first_records():
+    """Get first records from fully joined MCS-EPC dataset.
+
+    Returns:
+        pd.DataFrame: MCS records joined with first EPC.
+    """
+    mcs_epc = get_mcs_epc_domestic()
+    regions = get_countries()
+
+    mcs_epc = mcs_epc.merge(regions, on="postcode", how="left")
+    mcs_epc = mcs_epc.loc[mcs_epc["country"] == "Wales"].reset_index(drop=True)
+
+    first_records = (
+        mcs_epc.sort_values("INSPECTION_DATE").groupby("original_mcs_index").head(1)
+    )
+
+    return first_records
+
+
+# PROCESSING JOINED MCS-EPC
+
+
+def add_hp_when_built_column(first_records):
+    """Add a column to a DataFrame indicating that the property had a HP when it was built.
+
+    Args:
+        first_records (pd.DataFrame): MCS records joined to first EPC records.
+
+    Returns:
+        pd.DataFrame: Dataset with added "assumed_hp_when_built" column.
+    """
+
+    first_records["diff_epc_to_mcs"] = (
+        first_records["commission_date"] - first_records["INSPECTION_DATE"]
+    ).dt.days
+
+    # Assume dwelling was built with HP if:
+    # - first EPC shows it as a new dwelling
+    # - time difference between EPC inspection when dwelling was built and HP installation is less than 1 year
+    first_records["assumed_hp_when_built"] = (
+        first_records["TRANSACTION_TYPE"] == "new dwelling"
+    ) & (first_records["diff_epc_to_mcs"] < 365)
```

Comment:
Just checked and <=365 doesn't change the results :)

There are a few like this - hard to know which ones are legit, I can imagine a self-builder getting a HP installed a year or so before getting the EPC done. In fact I don't even know if self-built properties need an EPC because they're not being sold! But there are only 12 records in Wales where the MCS record is more than a year before the EPC assessment